src/backend/cdb/cdbmutate.c

```diff
@@ -2610,12 +2619,112 @@ shareinput_mutator_xslice_4(Node *node, PlannerInfo *root, bool fPop)
 	return true;
 }
 
+static bool
+root_slice_is_executed_on_coordinator(Plan *plan, PlannerInfo *root)
+{
+	Query	   *query = root->parse;
+
+	if (query->parentStmtType)
+	{
+		/*
+		 * For CTAS, SELECT INTO, COPY INTO, REFRESH MATERIALIZED VIEW, the
+		 * root slice is not executed on the coordinator.
+		 */
+		return false;
+	}
+	else if (IsA(plan, ModifyTable))
+	{
+		ModifyTable *mt = (ModifyTable *) plan;
+
+		if (list_length(mt->resultRelations) > 0)
+		{
+			ListCell   *lc = list_head(mt->resultRelations);
+			Oid			reloid = getrelid(lfirst_int(lc), query->rtable);
+
+			if (GpPolicyFetch(reloid)->ptype != POLICYTYPE_ENTRY)
+			{
+				/*
+				 * For the modification node by Postgres planner, if the table
+				 * distribution policy is not master-only, then the root slice
+				 * is not executed on the coordinator.
+				 */
+				return false;
+			}
+		}
+	}
+	else if (IsA(plan, DML))
+	{
+		DML		   *dml = (DML *) plan;
+		Oid			reloid = getrelid(dml->scanrelid, query->rtable);
+
+		if (GpPolicyFetch(reloid)->ptype != POLICYTYPE_ENTRY)
+		{
+			/*
+			 * For the modification node by ORCA planner, if the table
+			 * distribution policy is not master-only, then the root slice is
+			 * not executed on the coordinator.
+			 */
+			return false;
+		}
+	}
+	else if (plan->flow)
+	{
+		Flow	   *flow = plan->flow;
+
+		if (root->glob->is_parallel_cursor)
+		{
+			if (flow->flotype == FLOW_SINGLETON &&
+				(flow->locustype == CdbLocusType_Entry ||
+				 flow->locustype == CdbLocusType_General ||
+				 flow->locustype == CdbLocusType_SingleQE))
+			{
+				/*
+				 * For these scenarios, parallel retrieve cursor needs to run
+				 * on coordinator, since endpoint QE needs to interact with
+				 * the retrieve connections.
+				 */
+				return true;
+			}
+			else
+			{
+				return false;
+			}
+		}
+
+		if (flow->flotype != FLOW_SINGLETON || flow->segindex >= 0)
+		{
+			/*
+			 * For non-singleton or singleton on segment, the root slice is
+			 * not executed on the coordinator.
+			 */
+			return false;
+		}
+	}
+
+	/*
+	 * In all other cases, the root slice is executed on the coordinator.
+	 */
+	return true;
+}
+
 Plan *
 apply_shareinput_xslice(Plan *plan, PlannerInfo *root)
 {
 	PlannerGlobal *glob = root->glob;
 	ApplyShareInputContext *ctxt = &glob->share;
 	ShareInputContext walker_ctxt;
+	Motion		fakeMotion = {0};
+	Plan		fakePlan = {0};
+	Flow		fakeFlow = {0};
+
+	fakeMotion.plan.lefttree = &fakePlan;
+	fakeMotion.plan.lefttree->flow = &fakeFlow;
+
+	if (root_slice_is_executed_on_coordinator(plan, root))
+	{
+		fakeFlow.flotype = FLOW_SINGLETON;
+		fakeFlow.segindex = -1;
+	}
```

Comment:
Where is the else condition? Do you rely on fact that first enum value is zero?

Reply:
> Where is the else condition? Do you rely on fact that first enum value is zero?

done